internal/config/incremental_sync.go

```diff
@@ -0,0 +1,256 @@
+	for _, t := range ts {
+		r.configChangeTimestamps[tableName] = t.GetChangedAt()
+
+		logger := r.logger.With(zap.String("table", tableName), zap.Inline(t))
+
+		if t.IsDeleted() {
+			countDel++
+			logger.Debug("Marking entry as deleted")
+			(*changeConfigSetField)[t.GetPrimaryKey()] = nil
+			continue
+		}
+
+		if t, ok := any(t).(IncrementalConfigurableInitAndValidatable); ok {
+			err := t.IncrementalInitAndValidate()
+			if err != nil {
+				countErr++
+				logger.Errorw("Cannot validate entry, skipping element", zap.Error(err))
+				continue
+			}
+		}
+
+		countLoad++
+		logger.Debug("Loaded entry")
+		(*changeConfigSetField)[t.GetPrimaryKey()] = t
+	}
+
+	stmtLogger = stmtLogger.With("took", time.Since(startTime))
+	if countDel > 0 || countErr > 0 || countLoad > 0 {
+		stmtLogger.Debugw("Fetched incremental configuration updates",
+			zap.Int("deleted-elements", countDel),
+			zap.Int("faulty-elements", countErr),
+			zap.Int("loaded-elements", countLoad))
+	} else {
+		stmtLogger.Debug("No configuration updates are available")
+	}
+
+	return nil
+}
+
+// errRemoveAndAddInstead is a special non-error which might be expected from incrementalApplyPending's updateFn to
+// signal that the current element should be updated by being deleted through the deleteFn first and added again by the
+// createFn hook function.
+var errRemoveAndAddInstead = errors.New("re-adding by invoking the deletion function followed by the creation function")
+
+// incrementalApplyPending merges the incremental change from RuntimeConfig.configChange into the main ConfigSet.
+//
+// The recently fetched incremental change can be of three different types:
+//   - Newly created elements. Therefore, the createFn callback function will be called upon it, allowing both further
+//     initialization and also aborting by returning an error.
+//   - Changed elements. The updateFn callback function receives the current and the updated element, expecting the
+//     implementation to synchronize the necessary changes into the current element. This hook is allowed to return an
+//     error as well. However, it might also return the special errRemoveAndAddInstead, resulting in the old element to
+//     be deleted first and then re-added, with optional calls to the other two callbacks included.
+//   - Finally, deleted elements. Additional cleanup might be performed by the deleteFn.
+//
+// If no specific callback action is necessary, each function can be nil. A nil updateFn results in the same behavior as
+// errRemoveAndAddInstead.
+func incrementalApplyPending[
+	BaseT any,
+	PK comparable,
+	T interface {
+		*BaseT
+		IncrementalConfigurable[PK]
+	},
+](
+	r *RuntimeConfig,
+	configSetField, changeConfigSetField *map[PK]T,
+	createFn func(newElement T) error,
+	updateFn func(curElement, update T) error,
+	deleteFn func(delElement T) error,
+) (
+	appliedChanges bool,
+) {
+	startTime := time.Now()
+	tableName := database.TableName(*new(T))
```

Comment:
Hm, that looks pretty confusing. If I didn't get something wrong, `*new(T)` should do exactly the same as `T(nil)` here.

Even if we don't change which of the type parameters is used throughout the function, maybe we should rename them to `T` (from `BaseT`) and `Tptr` (from `T`) so that it's at least somewhat clear that one of them is actually a pointer type. At least `*new(Tptr)` would look more suspect to me that it currently does.

Reply:
Technically, `new(BaseT)` should also work. However, I found this confusing as otherwise only `T` is being used. I also disliked the idea of introducing a new variable to be used just here. Furthermore, calling `new` is useless anyway. Somehow, my brain did not end up at `T(nil)`, which I now used.

I started to take a liking for the current generic type names, being constructed from somehow descriptive names to the good old `T`, indicating that this is the generic type to use.